ipykernel/iostream.py

```diff
@@ -105,87 +109,183 @@ def _check_mp_mode(self):
                 self._setup_pipe_out()
             return CHILD
 
-    def set_parent(self, parent):
-        self.parent_header = extract_header(parent)
-
+    
+    def start(self):
+        """Start the IOPub thread"""
+        self.thread.start()
+    
+    def stop(self):
+        """Stop the IOPub thread"""
+        self.io_loop.add_callback(self.io_loop.stop)
+        self.thread.join()
+    
     def close(self):
-        self.pub_socket = None
+        self.socket.close()
+        self.socket = None
 
     @property
     def closed(self):
-        return self.pub_socket is None
-
-    def _flush_from_subprocesses(self):
-        """flush possible pub data from subprocesses into my buffer"""
-        if not self._pipe_flag or not self._is_master_process():
-            return
-        for i in range(self._subprocess_flush_limit):
-            if self._pipe_poller.poll(0):
-                msg = self._pipe_in.recv_multipart()
-                if msg[0] != self._pipe_uuid:
-                    continue
-                else:
-                    self._buffer.write(msg[1].decode(self.encoding, 'replace'))
-                    # this always means a flush,
-                    # so reset our timer
-                    self._start = 0
-            else:
-                break
+        return self.socket is None
+    
+    def send_multipart(self, *args, **kwargs):
+        """send_multipart schedules actual zmq send in my thread.
+        
+        If my thread isn't running (e.g. forked process), send immediately.
+        """
 
-    def _schedule_flush(self):
-        """schedule a flush in the main thread
+        if self.thread.is_alive():
+            self.io_loop.add_callback(lambda : self._really_send(*args, **kwargs))
+        else:
+            self._really_send(*args, **kwargs)
+    
+    def _really_send(self, msg, *args, **kwargs):
+        """The callback that actually sends messages"""
+        mp_mode = self._check_mp_mode()
 
-        only works with a tornado/pyzmq eventloop running
-        """
-        if IOLoop.initialized():
-            IOLoop.instance().add_callback(self.flush)
+        if mp_mode != CHILD:
+            # we are master, do a regular send
+            self.socket.send_multipart(msg, *args, **kwargs)
+        else:
+            # we are a child, pipe to master
+            kwargs['copy'] = False
+            kwargs['track'] = True
+            tracker = self._pipe_out.send_multipart([self._pipe_uuid] + msg, *args, **kwargs)
+            try:
+                tracker.wait(1)
+            except Exception as e:
+                print("Failed to send: %s" % e, file=sys.__stderr__)
+                pass
+
+
+class BackgroundSocket(object):
+    """Wrapper around IOPub thread that provides zmq Socket API"""
+    io_thread = None
+    
+    def __init__(self, io_thread):
+        self.io_thread = io_thread
+    
+    def __getattr__(self, attr):
+        return getattr(self.io_thread.socket, attr)
```

Comment:
When I'm overriding `__getattr__` and `__setattr__`, I like to check for attempts at `__special__` attribute access, and make them behave as normal; otherwise things can get quite confused.